File: 4/4.js
// # A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// # Find the largest palindrome made from the product of two 3-digit numbers.

function findPalisUpTo(integer){
  var palis = [];
  var paliDigits = [];
  var potentialPalis =  [];
  for (i = 100001; i < integer; i++){
    paliDigits = i.toString().split('');
    if ((paliDigits[0] === paliDigits[5]) && (paliDigits[1] === paliDigits[4]) && (paliDigits[2] === paliDigits[3])){
      palis.push(i);
    };
  };
  return palis
};

function findFactors(integer){
  factorArray = [];
  for (k = 2; k < integer; k++){
    if (integer % k === 0){
      factorArray.push(k);
    }
  };
  return factorArray
};

function findLargestPaliWithFactors(){
  paliFactors = [];
  palis = findPalisUpTo(999999);
  for (z = 1; z < palis.length; z++){
    paliFactors = findFactors(palis[palis.length - z])
      for (j = 0; j < paliFactors.length; j++){
        if ((paliFactors[j].toString().length === 3) && (paliFactors[j+1].toString().length === 3) && (paliFactors[j] * paliFactors[j + 1] === palis[palis.length - z])){
          return palis[palis.length - z]
        };
      };
  };
};

findLargestPaliWithFactors();
// =>906609

File: 4/4.rb
# A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product of two 3-digit numbers.

def find_palindromes_up_to(integer)
  palis = []
  pali_digits = []
  potential_palis = (0..integer).to_a
  potential_palis.each do |pali|
    pali_digits = pali.to_s.split('')
    if (pali_digits[0] == pali_digits[5]) && (pali_digits[1] == pali_digits[4]) && (pali_digits[2] == pali_digits[3])
      palis << pali
    end
  end
  palis
end

def find_factors(integer)
  factor_array = []
  potential_factor_array = (2..integer).to_a
  potential_factor_array.each do |factor|
    if integer % factor == 0
      factor_array << factor
    end
  end
  return factor_array
end

def find_largest_pali_with_factors
  pali_factors = []
  palis = find_palindromes_up_to(999 * 999)
  palis.reverse.each do |pali|
    pali_factors = find_factors(pali)
    if pali_factors.length >= 1
        pali_factors.each_with_index do |factor, index|
          if (factor.to_s.size == 3) && (pali_factors[index+1].to_s.size == 3) && (factor * pali_factors[index+1] == pali)
            return  p pali
          end
        end
    end
  end
end

find_largest_pali_with_factors
# => 906609
